mixer/pkg/runtime/config/ephemeral.go

```diff
@@ -333,68 +515,161 @@ func (e *Ephemeral) processRuleConfigs(
 		}
 
 		// extract the set of actions from the rule, and the handlers they reference.
-		actions := make([]*ActionStatic, 0, len(cfg.Actions))
+		// A rule can have both static and dynamic actions.
+
+		actionsStat := make([]*ActionStatic, 0, len(cfg.Actions))
+		actionsDynamic := make([]*ActionDynamic, 0, len(cfg.Actions))
 		for i, a := range cfg.Actions {
 			log.Debugf("Processing action: %s[%d]", ruleName, i)
+			var isHandlerStat bool
+			var isHandlerDynamic bool
+			var sahandler *HandlerStatic
+			var dahandler *HandlerDynamic
+			hdlName, hdlAltName := canonicalize(a.Handler, constant.HandlerKind, ruleKey.Namespace)
+
+			handlerName := hdlName
+			if sahandler, isHandlerStat = sHandlers[hdlName]; !isHandlerStat {
+				handlerName = hdlAltName
+				sahandler, isHandlerStat = sHandlers[hdlAltName]
+			}
 
-			var found bool
-			var handler *HandlerStatic
-			handlerName := canonicalize(a.Handler, ruleKey.Namespace)
-			if handler, found = handlers[handlerName]; !found {
-				appendErr(errs, fmt.Sprintf("action='%s[%d]'", ruleName, i), counters.ruleConfigError, "Handler not found: handler='%s'",
-					handlerName)
+			if !isHandlerStat {
+				handlerName = hdlName
+				if dahandler, isHandlerDynamic = dHandlers[hdlName]; !isHandlerDynamic {
+					handlerName = hdlAltName
+					dahandler, isHandlerDynamic = dHandlers[hdlAltName]
+				}
+			}
+
+			if !isHandlerStat && !isHandlerDynamic {
+				appendErr(errs, fmt.Sprintf("action='%s[%d]'", ruleName, i), counters.ruleConfigError,
+					"Handler not found: handler='%s'", a.Handler)
 				continue
 			}
 
-			// Keep track of unique instances, to avoid using the same instance multiple times within the same
-			// action
-			uniqueInstances := make(map[string]bool, len(a.Instances))
+			if isHandlerStat {
+				// Keep track of unique instances, to avoid using the same instance multiple times within the same
+				// action
+				uniqueInstances := make(map[string]bool, len(a.Instances))
+
+				actionInstances := make([]*InstanceStatic, 0, len(a.Instances))
+				for _, instanceNameRef := range a.Instances {
+					in1, in2 := canonicalize(instanceNameRef, constant.InstanceKind, ruleKey.Namespace)
```

Comment:
This logic of trying first and alternate is repeated 4 times. wonder if it can be abstracted as
`getCanonicalRef(4 canonicalize_args, lookup1 func (key) interface, lookup2 func (key) interface) interface{}`